agents/ten_packages/extension/openai_chatgpt_python/extension.py

```diff
@@ -0,0 +1,326 @@
+                output_data.set_property_bool(DATA_OUT_TEXT_DATA_PROPERTY_TEXT_END_OF_SEGMENT, True)
+                ten_env.send_data(output_data)
+                logger.info(f"Greeting [{greeting}] sent")
+            except Exception as err:
+                logger.info(f"Failed to send greeting [{greeting}]: {err}")
+        ten_env.on_start_done()
+
+    def on_stop(self, ten_env: TenEnv) -> None:
+        logger.info("on_stop")
+
+        # TODO: clean up resources
+
+        ten_env.on_stop_done()
+
+    def on_deinit(self, ten_env: TenEnv) -> None:
+        logger.info("on_deinit")
+        ten_env.on_deinit_done()
+
+    def on_cmd(self, ten_env: TenEnv, cmd: Cmd) -> None:
+        logger.info(f"on_cmd json: {cmd.to_json()}")
+
+        cmd_name = cmd.get_name()
+
+        if cmd_name == CMD_IN_FLUSH:
+            asyncio.run_coroutine_threadsafe(self._flush_queue(), self.loop)
+            ten_env.send_cmd(Cmd.create(CMD_OUT_FLUSH), None)
+            logger.info("on_cmd sent flush")
+            status_code, detail = StatusCode.OK, "success"
+        else:
+            logger.info(f"on_cmd unknown cmd: {cmd_name}")
+            status_code, detail = StatusCode.ERROR, "unknown cmd"
+        
+        cmd_result = CmdResult.create(status_code)
+        cmd_result.set_property_string("detail", detail)
+        ten_env.return_result(cmd_result, cmd)
+
+    def on_data(self, ten_env: TenEnv, data: Data) -> None:
+        # Get the necessary properties
+        is_final = get_property_bool(data, DATA_IN_TEXT_DATA_PROPERTY_IS_FINAL)
+        input_text = get_property_string(data, DATA_IN_TEXT_DATA_PROPERTY_TEXT)
+
+        if not is_final:
+            logger.info("ignore non-final input")
+            return
+        if not input_text:
+            logger.info("ignore empty text")
+            return
+
+        logger.info(f"OnData input text: [{input_text}]")
+
+        # Start an asynchronous task for handling chat completion
+        asyncio.run_coroutine_threadsafe(self.queue.put(input_text), self.loop)
+
+    def on_audio_frame(self, ten_env: TenEnv, audio_frame: AudioFrame) -> None:
+        # TODO: process pcm frame
+        pass
+
+    def on_video_frame(self, ten_env: TenEnv, video_frame: VideoFrame) -> None:
+        # logger.info(f"OpenAIChatGPTExtension on_video_frame {frame.get_width()} {frame.get_height()}")
+        self.image_data = video_frame.get_buf()
+        self.image_width = video_frame.get_width()
+        self.image_height = video_frame.get_height()
+        return
+
+    async def _process_queue(self, ten_env: TenEnv):
+        """Asynchronously process queue items one by one."""
+        while True:
+            # Wait for an item to be available in the queue
+            message = await self.queue.get()
+            try:
+                # Create a new task for the new message
+                start_time = get_current_time()
```

Comment:
It might be too late to set the timestamp here, move it to `on_data` seems better.